### PR TITLE
Support for linear gradients: add additional type to waveColor and progressColor parameters

### PR DESCRIPTION
**Pull request by califken**
### Short description of changes:
Added feature to pass an array of colors into the waveColor and progressColor parameters.  Edited the setFillStyles function to check if the provided argument is a string or an object.  In the case of an object, a linear gradient is created, with each item in the array added as a color stop.

wavecolorgradient
wavecolorgradientcodeexample

### Breaking in the external API:
Not sure.

### Breaking changes in the internal API:
None, as it runs the check to see if it is a string, therefore should be backwards compatible.

### Todos/Notes:
If this feature is integrated, I would eventually like to integrate the ability to pass images and even a video as the wave background.

### CODE REVIEW

**Review by thijstriemstra — CHANGES_REQUESTED** [review]
That's great @califken! Can you also add an example that showcases this feature? And it'll also need a changelog entry? And make sure to run `npm run lint` because there's a lint error: https://github.com/katspaugh/wavesurfer.js/pull/2345/checks?check_run_id=3384740565#step:8:26

**Comment by califken** [comment]
> That's great @califken!
> 
> * Can you also add an example that showcases this feature?
Done, added a gradient-fill-styles directory in the examples folder.  Not sure if there is an index that I should add it to somewhere.
> * And it'll also need a changelog entry?
Done, added a changelog entry to CHANGES.md.  Bumped the version up to 5.3.0.
> * And make sure to run `npm run lint` because there's a lint error: https://github.com/katspaugh/wavesurfer.js/pull/2345/checks?check_run_id=3384740565#step:8:26
Corrected the issue.
> * add a unit test
Added @test {WaveSurfer#setWaveColorGradient} to wavesurfer.spec.js
> Also, what does it look like with bars (#1722)?

It looks awesome with bars!  :)
Screen Shot 2021-08-20 at 3 41 35 PM

**Comment by califken** [comment]
Rebased and pushed the one commit.  Ready to go.

**Review by thijstriemstra — CHANGES_REQUESTED** [review]

File: CHANGES.md [hunk 1]
```diff
@@ -1,5 +1,8 @@
 wavesurfer.js changelog
 =======================
+5.3.0 (20.08.2021)
```
[the remark this answers is not in the document]
Comment (califken) [reply CHANGES.md]: changed to unreleased

File: example/gradient-fill-styles/main.js [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    // Set the playhead to halfway through the media, as to demonstrate the colorProgress gradient
+    wavesurfer.on('ready', () => wavesurfer.seekTo(.5));
+
+    // Play button
+    const button = document.querySelector('[data-action="play"]');
+
+    button.addEventListener('click', wavesurfer.playPause.bind(wavesurfer));
+
+
+    // WaveSurfer with bars example
```
[the remark this answers is not in the document]
Comment (califken) [reply example/gradient-fill-styles/main.js]: updated comment

File: package.json [hunk 3]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "wavesurfer.js",
-  "version": "5.2.0",
+  "version": "5.3.0",
```
[the remark this answers is not in the document]
Comment (califken) [reply package.json]: changed back to 5.2.0

File: example/gradient-fill-styles/index.html [hunk 4]
```diff
@@ -0,0 +1,180 @@
+<!DOCTYPE html>
+<html>
+
+<head>
+    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
+    <title>wavesurfer.js | Waveform using bars</title>
```
[the remark this answers is not in the document]
Comment (califken) [reply example/gradient-fill-styles/index.html]: changed title to wavesurfer.js | Gradient Fill Styles Example

File: src/wavesurfer.js [hunk 5]
```diff
@@ -1154,8 +1155,9 @@ export default class WaveSurfer extends util.Observer {
     /**
      * Set the fill color of the waveform behind the cursor.
      *
-     * @param {string} color A CSS color string.
+     * @param {string|object} color A CSS color string, or an array of CSS color strings.
      * @example wavesurfer.setProgressColor('#400');
+     * @example wavesurfer.setWavecolor(['blue','black','yellow']);
```
Comment (thijstriemstra) [inline src/wavesurfer.js:1160]: setWaveColor example doesn't belong here.

File: src/drawer.canvasentry.js [hunk 6]
```diff
@@ -146,14 +146,25 @@ export default class CanvasEntry {
...
+        if (typeof waveColor == 'string') {
+            this.waveCtx.fillStyle = waveColor;
+        } else if (typeof waveColor == 'object') {
+            const waveGradient = this.waveCtx.createLinearGradient(0, 0, 0, this.waveCtx.canvas.height);
+            waveColor.forEach((color, index) => waveGradient.addColorStop((index / waveColor.length), color));
+            this.waveCtx.fillStyle = waveGradient;
+        }
         if (this.hasProgressCanvas) {
-            this.progressCtx.fillStyle = progressColor;
+            if (typeof progressColor == 'string') {
```
Comment (msuhov) [inline src/drawer.canvasentry.js:161]: it looks like a duplicate of previous code, I'd rather extract it to an utility or something like this:
```
this.waveCtx.fillStyle = getFillStyle(this.waveCtx, waveColor)

if (this.hasProgressCanvas) {
  this.progressCtx.fillStyle = getFillStyle(this.progressCtx, progressColor)
}


...

export const getFillStyle = (context, color) => {
  if (typeof waveColor == 'string') {
    return color
  }

  const waveGradient = context.createLinearGradient(0, 0, 0, context.canvas.height);
  color.forEach((value, index) => waveGradient.addColorStop((index / color.length), value));

  return waveGradient
}
```

Comment (califken) [reply src/drawer.canvasentry.js]: Agreed, included in latest commit.

File: src/wavesurfer.js (hunk 5 above)
Comment (califken) [reply src/wavesurfer.js]: Agreed, removed in latest commit.

File: spec/wavesurfer.spec.js [hunk 7]
```diff
@@ -264,6 +264,27 @@ describe('WaveSurfer/playback:', function() {
...
+            "red",
+            "green",
+            "purple",
+            "yellow",
+            "rgba(0,255,255,.5)"
+        ];
+        wavesurfer.setWaveColor(colors);
+        const waveColor = wavesurfer.getWaveColor();
+
+        expect(waveColor).toEqual([
```
Comment (thijstriemstra) [inline spec/wavesurfer.spec.js:279]: `toEqual(colors)`?

File: src/drawer.canvasentry.js [hunk 8]
```diff
@@ -146,17 +146,35 @@ export default class CanvasEntry {
     /**
      * Set the fill styles for wave and progress
      *
-     * @param {string} waveColor Fill color for the wave canvas
-     * @param {?string} progressColor Fill color for the progress canvas
+     * @param {string|string[]} waveColor Fill color for the wave canvas, or an array of colors to apply as a gradient
```
Comment (thijstriemstra) [inline src/drawer.canvasentry.js:149]: spread sentence over 2 lines

File: src/drawer.canvasentry.js [hunk 9]
```diff
@@ -146,17 +146,35 @@ export default class CanvasEntry {
...
+     * @param {?string|string[]} progressColor Fill color for the progress canvas, or an array of colors to apply as a gradient
```
Comment (thijstriemstra) [inline src/drawer.canvasentry.js:150]: spread sentence over 2 lines

File: src/drawer.canvasentry.js [hunk 10]
```diff
@@ -146,17 +146,35 @@ export default class CanvasEntry {
...
+            this.progressCtx.fillStyle = this.getFillStyle(this.progressCtx, progressColor);
         }
     }
 
+    /**
+     * Set the fill styles for wave and progress
+     *
+     * @param {CanvasRenderingContext2D} ctx Rendering context of target canvas
+     * @param {string|string[]} color Fill color for the wave canvas, or an array of colors to apply as a gradient
+     * @returns {string | CanvasGradient} Returns a string fillstyle value, or a canvas gradient
```
Comment (thijstriemstra) [inline src/drawer.canvasentry.js:165]: add version marker for 5.3.0 to method (https://jsdoc.app/tags-since.html)

**Review by thijstriemstra — CHANGES_REQUESTED** [review]
Build is also failing.

File: src/drawer.canvasentry.js (hunk 10 above)
Comment (thijstriemstra) [reply src/drawer.canvasentry.js:165]: `@returns {string | CanvasGradient}`: remove extra spaces

**Review by thijstriemstra — APPROVED** [review]
thx!

**Comment by jadenmazzone** [comment]
@thijstriemstra is it possible to have the linear gradient show left to right - photo attached to see what I mean
Screen Shot 2022-12-07 at 11 20 19 AM

**Pull request merged**